File: Data_Structure/python/4949.py
dic = {
  ")":"(",
  "}":"{",
  "]":"[",
}
open_parentheses = ["(","[","{"]
close_parentheses = [")","]","}"]
result = ""
while True:
  stack = []
  s = input()
  if s ==".":
    break
  check = "yes"
  for char in s:
    if char in open_parentheses:
      stack.append(char)
      continue
    if char in close_parentheses :
      if not stack or dic[char]!= stack[-1]:
        check = "no"
        break
      stack.pop()
  if stack:
    check = "no"
  result += check + "\n"
print(result)

File: Data_Structure/javascript/4949.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');

result = "";
let dic = new Map();
dic.set(")","(");
dic.set("]","[");
dic.set("}","{");

for (let i = 0 ; i<input.length-2 ; i++){
  stack = [];
  let char;
  for (let j = 0 ; j<input[i].length ; j++){
    char = input[i][j]; 
    s = "yes";
    if (char === "(" || char === "}" || char === "[") {
      stack.push(char);
      continue
    }
    if (char === ")" || char === "}" || char === "]") {
      if (stack.length === 0){
        s = "no";
        break;
      }
      if (stack.pop() !== dic.get(char)){
        s = "no";
        break;
      }
    }
  }
  if (stack.length !== 0) s = "no";
  result += s;
  if (i !== input.length-3) result += "\n";
}
console.log(result);